### PR TITLE
[Kernels]  add inference token attention kernel

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_kernels/triton/test_token_attn_1.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import math
+
+import torch
+
+from colossalai.kernel.triton.token_attention_kernel import token_attn_fwd_1
+
+
+def torch_attn(xq, xk, bs, seqlen, num_head, head_dim):
```
Comment (tiandiao123) [inline tests/test_kernels/triton/test_token_attn_1.py:8]: tests follows this format? https://github.com/hpcaitech/ColossalAI/pull/4485/files#diff-232c19b679c915358eb4e7a5066a4110c9b2c7a7aca467597966e4d6fe8e9818R23

**Review by tiandiao123 — CHANGES_REQUESTED** [review]

File: tests/test_kernels/triton/test_token_attn_1.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    xq = xq.view(bs, 1, num_head, head_dim)
+    xk = xk.view(bs, seqlen, num_head, head_dim)
+    keys = xk
+    xq = xq.transpose(1, 2)
+    keys = keys.transpose(1, 2)
+    scores = (torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(head_dim)).squeeze().transpose(0, 1).reshape(
+        num_head, -1)
+    # print("s  ", scores.shape)
```
Comment (tiandiao123) [inline tests/test_kernels/triton/test_token_attn_1.py:16]: deleted this?

File: colossalai/kernel/triton/token_attention_kernel.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+            attn_out.stride(2),
+            HEAD_DIM=dim,
+            BLOCK_N=BLOCK,
+            num_warps=num_warps,
+            num_stages=1,
+        )
+        return
+
+    @torch.no_grad()
+    def token_attention_fwd(q,
```
Comment (tiandiao123) [inline colossalai/kernel/triton/token_attention_kernel.py:299]: is this decoder function able to be used in bloom and llama?

**Review by tiandiao123 — CHANGES_REQUESTED** [review]

File: tests/test_kernels/triton/test_token_attn_1.py (hunk 2 above)
Comment (Xu-Kai) [reply tests/test_kernels/triton/test_token_attn_1.py]: Done.

File: tests/test_kernels/triton/test_token_attn_1.py (hunk 1 above)
Comment (Xu-Kai) [reply tests/test_kernels/triton/test_token_attn_1.py:20]: done.

File: colossalai/kernel/triton/token_attention_kernel.py (hunk 3 above)
Comment (Xu-Kai) [reply colossalai/kernel/triton/token_attention_kernel.py:299]: Yes.

**Review by tiandiao123 — APPROVED** [review]

**Review by tiandiao123 — APPROVED** [review]

**Pull request merged**